src/processes.c

```diff
@@ -1001,6 +923,126 @@ static procstat_t *ps_read_io (long pid, procstat_t *ps)
 	return (ps);
 } /* procstat_t *ps_read_io */
 
+static int ps_iter_dir (const char *dirname, const char *iterfile, void (*callback)(void *, char *, int), void *ctx)
+{
+	DIR           *dh;
+	struct dirent *ent;
+	char           filename[64];
+	FILE          *fh;
+	char           buffer[1024];
+
+	if ((dh = opendir (dirname)) == NULL)
+	{
+		DEBUG ("Failed to open directory `%s'", dirname);
+		return (-1);
+	}
+
+	while ((ent = readdir (dh)) != NULL)
+	{
+		if (ent->d_name[0] == '.')
+			continue;
+
+		ssnprintf (filename, sizeof (filename), "%s/%s/%s", \
+				dirname, ent->d_name, iterfile);
+		if ((fh = fopen (filename, "r")) == NULL)
+		{
+			DEBUG ("Failed to open file `%s'", filename);
+			continue;
+		}
+
+		while (fgets (buffer, sizeof(buffer), fh) != NULL)
+		{
+			callback(ctx, buffer, sizeof(buffer));
+		}
+
+		if (fclose (fh))
+		{
+			char errbuf[1024];
+			WARNING ("processes: fclose: %s",
+				sstrerror (errno, errbuf, sizeof (errbuf)));
+		}
+	}
+
+	closedir (dh);
+	return (0);
+} /* int ps_iter_dir */
+
+typedef struct taskstate_counter_s
+{
+	int running;
+	int sleeping;
+	int zombies;
+	int stopped;
+	int paging;
+	int blocked;
+} taskstate_counter_t;
+
+static void ps_state_callback (void *ctx, char *buffer, int buflen)
+{
+	taskstate_counter_t *counter;
+	char *fields[2];
+	int   numfields;
+
+	counter = (taskstate_counter_t *) ctx;
+
+	if (strncmp (buffer, "State:", 6) != 0)
+		return;
+
+	numfields = strsplit (buffer, fields, STATIC_ARRAY_SIZE (fields));
+	if (numfields < 2)
+		return;
+
+	switch (fields[1][0])
+	{
+		case 'R': counter->running++;  break;
+		case 'S': counter->sleeping++; break;
+		case 'D': counter->blocked++;  break;
+		case 'Z': counter->zombies++;  break;
+		case 'T': counter->stopped++;  break;
+		case 'W': counter->paging++;   break;
+	}
+}
+
+typedef struct taskswitch_counter_s
+{
+	derive_t cswitch_vol;
+	derive_t cswitch_invol;
+} taskswitch_counter_t;
+
+static void ps_ctxt_switch_callback (void *ctx, char *buffer, int buflen)
+{
+	taskswitch_counter_t *counter;
+	char    *fields[8];
+	int      numfields;
+	derive_t tmp;
+	char    *endptr;
+
+	counter = (taskswitch_counter_t *) ctx;
+
+	if (strncmp (buffer, "voluntary_ctxt_switches", 23) != 0
+		&& strncmp (buffer, "nonvoluntary_ctxt_switches", 26) != 0)
+		return;
+
+	numfields = strsplit (buffer, fields, STATIC_ARRAY_SIZE (fields));
+	if (numfields < 2)
+		return;
+
+	errno = 0;
+	endptr = NULL;
+	tmp = (derive_t) strtoll (fields[1], &endptr, /* base = */ 10);
```

Comment:
Please use `parse_value (fields[1], &value, DS_TYPE_DERIVE)` instead of `strtoll()`.